qmplay2: 24.04.07 -> 24.06.16

AndersonTorres opened the pull request:
## Description of changes


Closes https://github.com/NixOS/nixpkgs/issues/320986

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

AndersonTorres committed:
qmplay2: prepare to by-name migration

- get rid of libsForQt5 indirection
- postFixup instead of postInstall
- meta.mainProgram

AndersonTorres committed:
qmplay2: migrate to by-name

AndersonTorres committed:
qmplay2: detach sources acquisition to `sources.nix`

Because it uses fetchSubmodules, and maybe in the future it will
become discouraged.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/4124

AndersonTorres commented:
@ofborg build pkgsMusl.qmplay2

h7x4 commented on pkgs/by-name/qm/qmplay2/sources.nix:
```diff
@@ -0,0 +1,21 @@
+{
+  fetchFromGitHub,
+}:
+
+{
+  qmplay2 = let
+    self = {
+      pname = "qmplay2";
+      version = "24.06.16";
+
+      src = fetchFromGitHub {
+        owner = "zaps166";
+        repo = "QMPlay2";
+        rev = self.version;
+        fetchSubmodules = true;
+        hash = "sha256-ECDEyF16GwIKn1raG7J56HFMYk4G6ezwAG2BEy8hUZs=";
+      };
+    };
+  in
+    self;
+}
```
The changes seem reasonable, but I'm having a hard time wrapping my head around the reasoning behind this block. Do you mind if I ask?

How does it help that the code is put into a seperate file if `fetchSubmodules` become discouraged or deprecated in the future?

And what is up with the weirdly recursive let block? Does it somehow make it overridable as a fixpoint expression from `pkgs.qmplay2`?

AndersonTorres replied:
> How does it help that the code is put into a seperate file if fetchSubmodules become discouraged or deprecated in the future?

As far as I rememeber, `fetchSubmodules` has some problems with reproducibility.
My intention is to push those submodules directly in a near future.

But this would generate a big block of code that is just a truckload of fetchers.
It clutters the already big package file.
So I detached the source acquisition to a separated file.

It indeed can help maintenance, since `package.nix` will deal with the build proper and `sources.nix` will deal with sources acquisition.

I have made a similar split before:

https://github.com/NixOS/nixpkgs/pull/310831
https://github.com/NixOS/nixpkgs/pull/319877
https://github.com/NixOS/nixpkgs/pull/322093

> And what is up with the weirdly recursive let block? 

I do not like `rec`. It caused too much problem before.
I prefer let-in, since it shows clearly the origin of each attribute.

h7x4 replied:
Sorry, I think I worded my questions a bit poorly. What I meant to ask was:

- I can see why you would want to split the source fetchers off the package once they become large, but is there any merit to doing it preemptively?

- Avoiding `rec` is fair, but why use recursion within the `let` block? Wouldn't it be easier to both read and write (and possibly evaluate) something like this instead?

```nix
qmplay2 = 
  let
    version = "24.06.16";
  in
  {
    pname = "qmplay2";
    inherit version;

    src = fetchFromGitHub {
      owner = "zaps166";
      repo = "QMPlay2";
      rev = version;
      fetchSubmodules = true;
      hash = "sha256-ECDEyF16GwIKn1raG7J56HFMYk4G6ezwAG2BEy8hUZs=";
    };
  };
```

AndersonTorres replied:
> I can see why you would want to split the source fetchers off the package once they become large, but is there any merit to doing it preemptively?

1. Yes, I think it does. Not a huge merit, but a non-negligible one. 
2. Indeed I got rid of it right now.

> Wouldn't it be easier to both read and write (and possibly evaluate) something like this instead?

This is a matter of personal taste. 
`let-self-in-self` is a bit more typical in Nixpkgs, and I believe it has a clearer intention than merely factoring equal terms. In other words, it is a `rec` without the problems of `rec`.

AndersonTorres committed:
qmplay2: set strictDeps

AndersonTorres committed:
qmplay2: 24.05.23 -> 24.06.16

h7x4 commented:
I'm unable to launch the program on nixos, it gives me `Cannot mix incompatible Qt library (6.7.1) with this library (6.7.2)`. Do you happen to know whether this is an environment error, or something with the package? Maybe because I'm running nixos stable?

AndersonTorres commented:
I am using the unstable, and here it is working fine.

Pandapip1 commented on pkgs/by-name/qm/qmplay2/package.nix:
```diff
@@ -0,0 +1,111 @@
+  vulkan-tools,
+  # Configurable options
+  qtVersion ? "6", # Can be 5 or 6
+  sources ? callPackage ./sources.nix { },
```
https://github.com/NixOS/nixpkgs/issues/309892#issuecomment-2212447383

AndersonTorres replied:
done

AndersonTorres committed:
qmplay2: fetch submodules directly (get rid of fetchSubmodules)

AndersonTorres committed:
qmplay2: 24.04.07 -> 24.05.23

Pandapip1 reviewed (approved):
Diff LGTM; have not built or tested functionality

Pandapip1 commented on pkgs/by-name/qm/qmplay2/package.nix:
```diff
@@ -0,0 +1,113 @@
+      files, Rayman 2 music and chiptunes. It contains YouTube and MyFreeMP3
+      browser.
+    '';
+    changelog = "https://github.com/zaps166/QMPlay2/releases/tag/${finalAttrs.version}";
```
Wait whoops let's not use version in the changelog.

```suggestion
    changelog = "https://github.com/zaps166/QMPlay2/releases";
```

AndersonTorres replied:
Why not?

Pandapip1 replied:
A couple reasons:

1. `overrideAttrs`-ing the version to something that isn't a tag (e.g. `<version>-unstable-<date>`) will result in an invalid URL
2. `.../releases/tag/<tag>` doesn't show a full changelog, but `.../releases` does.

AndersonTorres replied:
I will link to https://github.com/zaps166/QMPlay2/blob/24.06.16/ChangeLog, then

AndersonTorres committed:
qmplay2: delete changelog link

Until a better idea for an unbreakable reference template appears.

(Maybe a local file ref?)

Pandapip1 commented on pkgs/by-name/qm/qmplay2/package.nix:
```diff
@@ -0,0 +1,113 @@
+in
+assert lib.elem qtVersion [ "5" "6" ];
+stdenv.mkDerivation (finalAttrs: {
+  pname = sources.qmplay2.pname + "-qt" + qtVersion;
```
Non-blocking nit: if I were writing this, I'd use
```suggestion
  pname = "${sources.qmplay2.pname}-qt${qtVersion}";
```

Pandapip1 commented on pkgs/by-name/qm/qmplay2/package.nix:
```diff
@@ -0,0 +1,113 @@
+      files, Rayman 2 music and chiptunes. It contains YouTube and MyFreeMP3
+      browser.
+    '';
+    changelog = "https://github.com/zaps166/QMPlay2/blob/${finalAttrs.src.rev}/ChangeLog";
```
Again, if `finalAttrs.src.rev` is overwritten to something like `refs/tags/<tag>`, this URL breaks.

CC @eclairevoyant and @fgaz as two people that I know also don't like this pattern.

eclairevoyant replied:
while true that it can break the url, it needs a discussion and a treewide anyway...I'd not block on this

AndersonTorres replied:
What about deleting all references to changelog treewide, then?

Pandapip1 commented on pkgs/by-name/qm/qmplay2/package.nix:
```diff
@@ -0,0 +1,113 @@
+
+  # Because we think it is better to use only lowercase letters!
+  # But sometimes we come across case-insensitive filesystems...
+  postFixup = ''
```
Is there a reason this is `postFixup` rather than `postInstall`?

AndersonTorres replied:
A long decade ago I thought it was not an installation thing.

Pandapip1 commented on pkgs/by-name/qm/qmplay2/package.nix:
```diff
@@ -0,0 +1,113 @@
+    description = "Qt-based Multimedia player";
+    longDescription = ''
+      QMPlay2 is a video and audio player. It can play all formats supported by
+      FFmpeg, libmodplug (including J2B and SFX). It also supports Audio CD, raw
```
```suggestion
      FFmpeg and libmodplug (including J2B and SFX). It also supports Audio CDs, raw
```

Pandapip1 commented on pkgs/by-name/qm/qmplay2/package.nix:
```diff
@@ -0,0 +1,113 @@
+    longDescription = ''
+      QMPlay2 is a video and audio player. It can play all formats supported by
+      FFmpeg, libmodplug (including J2B and SFX). It also supports Audio CD, raw
+      files, Rayman 2 music and chiptunes. It contains YouTube and MyFreeMP3
```
```suggestion
      files, Rayman 2 music and chiptunes. It also contains YouTube and MyFreeMP3
```

Pandapip1 commented on pkgs/by-name/qm/qmplay2/package.nix:
```diff
@@ -0,0 +1,113 @@
+    homepage = "https://github.com/zaps166/QMPlay2/";
+    description = "Qt-based Multimedia player";
+    longDescription = ''
+      QMPlay2 is a video and audio player. It can play all formats supported by
+      FFmpeg, libmodplug (including J2B and SFX). It also supports Audio CD, raw
+      files, Rayman 2 music and chiptunes. It contains YouTube and MyFreeMP3
+      browser.
```
```suggestion
      QMPlay2 is a video and audio player. It can play all formats supported by
      FFmpeg and libmodplug (including J2B and SFX). It also supports Audio CDs, raw
      files, Rayman 2 music, and chiptunes. It contains YouTube and MyFreeMP3
      browsers.
```
This long description is still not great, but at least it's grammatically correct.

eclairevoyant replied:
On a 6 week old PR, I'd rather not nit grammar, and just focus on the CI, expression, and runtime behaviour itself. A quick PR to clean up the desc wouldn't be unwelcome if needed.

AndersonTorres replied:
This was a copy-paste from upstream IIRC.

eclairevoyant commented:
Please rebase on current `master`, then as the CI mentions:

```
Please run the following in `nix-shell`:
nixfmt 'pkgs/by-name/qm/qmplay2/package.nix' 'pkgs/by-name/qm/qmplay2/sources.nix'
```

AndersonTorres committed:
qmplay2: nixfmt-rfc-style

(Can someone explain what happened with buildInputs??)

AndersonTorres committed:
qmplay2: create symlink at postInstall

AndersonTorres committed:
qmplay2: fix longDescription

Grammar and punctuation.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-already-reviewed/2617/1874
